# Segmenting and Clustering Neighborhoods in Toronto 
### Author: Rishabh Dingliwal
<b> This notebook contains the solutions for Coursera IBM Applied Data Science Capstone Week 3 Final Assignment

## Part 1

<b> Importing the required libraries </b>

In [1]:
import pandas as pd
import requests

<b> Connecting to the Wiki webpage and requesting data </b>

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
wiki_url = requests.get(url)
wiki_url

<Response [200]>

<b> Extracting the tables using pandas </b>

In [3]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [4]:
len(wiki_data), type(wiki_data)

(3, list)

<b> The extracted data is in the form of a list with 3 values - each of which is a table. We only need the first table, so dropping the last 2 tables </b>

In [5]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

<b> Dropping the rows which have 'Borough' as 'Not assigned' </b>

In [6]:
df = wiki_data[wiki_data['Borough'] != 'Not assigned']
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

<b> Grouping the dataframe by 'Postal Code' so that neighbourhoods with same postal codes are combined together into a single row </b>

In [7]:
df.groupby(by = 'Postal Code')
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

<b> Counting the number of rows having 'Neighbourhood' as 'Not assigned' </b> 

In [8]:
df.Neighbourhood.str.count('Not assigned').sum()

0

<b> Zero rows have a the value 'Not assigned' for the column 'Neighbourhood', so no need to drop anymore rows </b> <br>
<b> Resetting the index of the dataframe </b>

In [9]:
df = df.reset_index(drop = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

<b> Printing out the shape of the dataframe </b>

In [10]:
df.shape

(103, 3)

<b> We can see that the dataframe has 103 rows and 3 colums </b>

## Part 2

<b> Instead of the geocoder library I will be using the given CSV file, as suggested in the assignment </b> <br>
<b> Loading the CSV file into a pandas dataframe </b>

In [11]:
latlong_data = pd.read_csv('https://cocl.us/Geospatial_data')
latlong_data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

<b> Checking the shapes of our two dataframes, to check compatibility for merging </b>

In [12]:
print('Shape of our Wiki data: ', df.shape)
print('Shape of our Latitude-Longitude data: ', latlong_data.shape)

Shape of our Wiki data:  (103, 3)
Shape of our Latitude-Longitude data:  (103, 3)


<b> Joining/merging the two dataframes </b>

In [13]:
df = pd.merge(df, latlong_data, on = 'Postal Code')
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

<b> Printing out the shape of the combined dataframe </b>

In [14]:
df.shape

(103, 5)

## Part 3

<b> Taking inspiration from the previous labs where we analysed the clusetrs of neighbourhoods in New York, we cluster based on similarities of the venues category using the K-Means algorithm and the Foursquare API  </b> 

<b> Downloading and Importing the required libraries </b>

In [15]:
!pip install folium

from geopy.geocoders import Nominatim
import folium

<b> Getting coordinates of Toronto </b>

In [16]:
address = 'Toronto, Ontario'

geolocater = Nominatim(user_agent = 'toronto_explorer')
location = geolocater.geocode(address)
latitude = location.latitude
longitude = location.longitude

print('The coordinates of Toronto are', latitude, longitude)

The coordinates of Toronto are 43.6534817 -79.3839347


<b> Visualizing the map of Toronto with the neighbourhoods marked </b>

In [17]:
map_toronto = folium.Map(location = [latitude, longitude], zoom_start = 11)
for latitude, longitude, borough, neighbourhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighbourhood']):
    label_string = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label_string, parse_html = True)
    folium.CircleMarker([latitude, longitude], radius = 5, popup = label, color = 'red', fill = True).add_to(map_toronto)
    
map_toronto

<b> Defining the Foursquare API credentials</b>

In [18]:
CLIENT_ID = 'XFNQJ2BXRCL4SSOWPTZXQ5UAMCYLZXOMRD3DMDVMQUWVQGBP'
CLIENT_SECRET = 'JI2XUJEWV0OECP1Q21RYSHIHUSGGJQQEC1CZBRPGZ3B22BUH'
VERSION = '20180605'

print('My Client ID:', CLIENT_ID)
print('My Client Secret:', CLIENT_SECRET)

My Client ID: XFNQJ2BXRCL4SSOWPTZXQ5UAMCYLZXOMRD3DMDVMQUWVQGBP
My Client Secret: JI2XUJEWV0OECP1Q21RYSHIHUSGGJQQEC1CZBRPGZ3B22BUH


<b> Creating a function to get venue categories in Toronto </b>

In [19]:
def getNearbyVenues(names, latitude, longitude, radius = 500):
    venues_list = []
    
    for name, lat, long in zip(names, latitude, longitude):
        print(name)
        
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(CLIENT_ID, CLIENT_SECRET, VERSION, lat, long, radius)
        results = requests.get(url).json()['response']['groups'][0]['items']
        venues_list.append([(name, lat, long, v['venue']['name'], v['venue']['categories'][0]['name']) for v in results])
    
    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 'Neighbourhood Latitude', 'Neighbourhood Longitude', 'Venue', 'Venue Category']
    
    return (nearby_venues)

<b> Finding the venues in Toronto for each neighbourhood </b>

In [20]:
venues_in_toronto = getNearbyVenues(df['Neighbourhood'], df['Latitude'], df['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

<b> Finding total number of venues </b>

In [21]:
venues_in_toronto.shape

(1314, 5)

<b> Printing the first 5 values of the total venues dataframe </b>

In [22]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1           Variety Store      Food & Drink Shop  
2  Victoria Village Arena           Hockey Arena  
3               Portugril  Portuguese Restaurant  
4             Tim Hortons            Coffee Shop

<b> Checking the values based on 'Neighbourhood' </b>

In [23]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                      Victoria Village   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1299  Mimico NW, The Queensway West, South of Bloor,...   
1300  Mimico NW, The Queensway West, South of Bloor,...   
1301  Mimico NW, The Queensway West, South of Bloor,...   
1302  Mimico NW, The Queensway West, South of Bloor,...   
1303  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656           Variety Store   
2                  43.725882               -79.315572  Victoria Village Arena   
3                  43.725882               -79.315572               Portugril   
4                  43.725882               -79.315572             Tim Hortons   
...                      ...                      ...                     ...   
1299               43.628841               -79.520999              Wingporium   
1300               43.628841               -79.520999        South St. Burger   
1301               43.628841               -79.520999               Dollarama   
1302               43.628841               -79.520999          Healthy Planet   
1303               43.628841               -79.520999    Artisano Bakery Café   

             Venue Category  
0                      Park  
1         Food & Drink Shop  
2              Hockey Arena  
3     Portuguese Restaurant  
4               Coffee Shop  
...                     ...  
1299            Wings Joint  
1300           Burger Joint  
1301         Discount Store  
1302        Supplement Shop  
1303                 Bakery  

[408 rows x 5 columns]

<b> Checking for maximum number of venue category </b>

In [24]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                     Toronto Dominion Centre, Design Exchange   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.718518               -79.464763   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
Airport Lounge                   43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.379817   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.689026               -79.453512   
Yoga Studio                      43.715383               -79.321558   

                                              Venue  
Venue Category                                       
Accessories Store               Ardene Shoes Outlet  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
Airport Lounge                        Porter Lounge  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                          The National Club  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[231 rows x 4 columns]

<b> There are 239 different types of venue categories, with 'Accessories Store' having the maximum number of venues </b>

<b> One Hot Encoding the categorical features </b>

In [25]:
toronto_venues_categorical = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix = '', prefix_sep = '')
toronto_venues_categorical

Accessories Store  Airport  Airport Food Court  Airport Gate  \
0                     0        0                   0             0   
1                     0        0                   0             0   
2                     0        0                   0             0   
3                     0        0                   0             0   
4                     0        0                   0             0   
...                 ...      ...                 ...           ...   
1309                  0        0                   0             0   
1310                  0        0                   0             0   
1311                  0        0                   0             0   
1312                  0        0                   0             0   
1313                  0        0                   0             0   

      Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                0                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    0   
...              ...              ...               ...                  ...   
1309               0                0                 0                    0   
1310               0                0                 0                    0   
1311               0                0                 0                    0   
1312               0                0                 0                    0   
1313               0                0                 0                    0   

      Antique Shop  Aquarium  ...  Train Station  Truck Stop  \
0                0         0  ...              0           0   
1                0         0  ...              0           0   
2                0         0  ...              0           0   
3                0         0  ...              0           0   
4                0         0  ...              0           0   
...            ...       ...  ...            ...         ...   
1309             0         0  ...              0           0   
1310             0         0  ...              0           0   
1311             0         0  ...              0           0   
1312             0         0  ...              0           0   
1313             0         0  ...              0           0   

      Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                                 0                 0                      0   
1                                 0                 0                      0   
2                                 0                 0                      0   
3                                 0                 0                      0   
4                                 0                 0                      0   
...                             ...               ...                    ...   
1309                              0                 0                      0   
1310                              0                 0                      0   
1311                              0                 0                      0   
1312                              0                 0                      0   
1313                              0                 0                      0   

      Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                   0         0            0              0            0  
1                   0         0            0              0            0  
2                   0         0            0              0            0  
3                   0         0            0              0            0  
4                   0         0            0              0            0  
...               ...   

<b> Adding neighbourhood to the encoded dataframe </b>

In [26]:
toronto_venues_categorical['Neighbourhood'] = venues_in_toronto['Neighbourhood']

new_columns = [toronto_venues_categorical.columns[-1]] + list(toronto_venues_categorical.columns[:-1])
toronto_venues_categorical = toronto_venues_categorical[new_columns]

toronto_venues_categorical

Neighbourhood  Accessories Store  \
0                                             Parkwoods                  0   
1                                             Parkwoods                  0   
2                                      Victoria Village                  0   
3                                      Victoria Village                  0   
4                                      Victoria Village                  0   
...                                                 ...                ...   
1309  Mimico NW, The Queensway West, South of Bloor,...                  0   
1310  Mimico NW, The Queensway West, South of Bloor,...                  0   
1311  Mimico NW, The Queensway West, South of Bloor,...                  0   
1312  Mimico NW, The Queensway West, South of Bloor,...                  0   
1313  Mimico NW, The Queensway West, South of Bloor,...                  0   

      Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0           0                   0             0               0   
1           0                   0             0               0   
2           0                   0             0               0   
3           0                   0             0               0   
4           0                   0             0               0   
...       ...                 ...           ...             ...   
1309        0                   0             0               0   
1310        0                   0             0               0   
1311        0                   0             0               0   
1312        0                   0             0               0   
1313        0                   0             0               0   

      Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                   0                 0                    0             0   
1                   0                 0                    0             0   
2                   0                 0                    0             0   
3                   0                 0                    0             0   
4                   0                 0                    0             0   
...               ...               ...                  ...           ...   
1309                0                 0                    0             0   
1310                0                 0                    0             0   
1311                0                 0                    0             0   
1312                0                 0                    0             0   
1313                0                 0                    0             0   

      ...  Train Station  Truck Stop  Vegetarian / Vegan Restaurant  \
0     ...              0           0                              0   
1     ...              0           0                              0   
2     ...              0           0                              0   
3     ...              0           0                              0   
4     ...              0           0                              0   
...   ...            ...         ...                            ...   
1309  ...              0           0                              0   
1310  ...              0           0                              0   
1311  ...              0           0                              0   
1312  ...              0           0                              0   
1313  ...              0           0                              0   

      Video Game Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  \
0                    0                      0                0         0   
1                    0                      0                0         0   
2                    0                      0                0         0   
3                    0                      0                0         0   
4                    0                      0                0         0   
...                ...                    ...         

<b> Grouping dataframe by 'Neighbourhood', calculate the mean venue categories in each neighbourhood </b>

In [27]:
toronto_grouped = toronto_venues_categorical.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Accessories Store  \
0                                         Agincourt                0.0   
1                            Alderwood, Long Branch                0.0   
2   Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                   Bayview Village                0.0   
4                 Bedford Park, Lawrence Manor East                0.0   
..                                              ...                ...   
92                      Willowdale, Willowdale West                0.0   
93                                           Woburn                0.0   
94                                 Woodbine Heights                0.0   
95                                  York Mills West                0.0   
96                         York Mills, Silver Hills                0.0   

    Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0       0.0                 0.0           0.0             0.0   
1       0.0                 0.0           0.0             0.0   
2       0.0                 0.0           0.0             0.0   
3       0.0                 0.0           0.0             0.0   
4       0.0                 0.0           0.0             0.0   
..      ...                 ...           ...             ...   
92      0.0                 0.0           0.0             0.0   
93      0.0                 0.0           0.0             0.0   
94      0.0                 0.0           0.0             0.0   
95      0.0                 0.0           0.0             0.0   
96      0.0                 0.0           0.0             0.0   

    Airport Service  Airport Terminal  American Restaurant  Antique Shop  ...  \
0               0.0               0.0             0.000000           0.0  ...   
1               0.0               0.0             0.000000           0.0  ...   
2               0.0               0.0             0.000000           0.0  ...   
3               0.0               0.0             0.000000           0.0  ...   
4               0.0               0.0             0.043478           0.0  ...   
..              ...               ...                  ...           ...  ...   
92              0.0               0.0             0.000000           0.0  ...   
93              0.0               0.0             0.000000           0.0  ...   
94              0.0               0.0             0.000000           0.0  ...   
95              0.0               0.0             0.000000           0.0  ...   
96              0.0               0.0             0.000000           0.0  ...   

    Train Station  Truck Stop  Vegetarian / Vegan Restaurant  \
0             0.0         0.0                            0.0   
1             0.0         0.0                            0.0   
2             0.0         0.0                            0.0   
3             0.0         0.0                            0.0   
4             0.0         0.0                            0.0   
..            ...         ...                            ...   
92            0.0         0.0                            0.0   
93            0.0         0.0                            0.0   
94            0.0         0.0                            0.0   
95            0.0         0.0                            0.0   
96            0.0         0.0                            0.0   

    Video Game Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  \
0                0.0                    0.0              0.0       0.0   
1                0.0                    0.0              0.0       0.0   
2                0.0                    0.0              0.0       0.0   
3                0.0                    0.0              0.0       0.0   
4                0.0                    0.0              0.0       0.0   
..               ...                    ...              ...       ...   
92               0.0                    0.0              0.0       0.0   
93               0.0              

<b> Creating a function to get the top most common venue categories </b>

In [28]:
def most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending = False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

<b> Importing required libraries </b>

In [29]:
import numpy as np

<b> There are too many venue categories. So we atke only top 10 to cluster the neighbourhoods </b>

In [30]:
num_top_venues = 10
indicators = ['st', 'nd', 'rd']

columns = ['Neighbourhood']
for ind in np.arange(num_top_venues): # Equivalent to for ind in range(0, num_top_venues)
    try:
        columns.append('{}{} Most Common Venue'.format(ind + 1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind + 1))

neighbourhoods_venues_sorted = pd.DataFrame(columns = columns)
neighbourhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted

Neighbourhood 1st Most Common Venue  \
0                                         Agincourt                Lounge   
1                            Alderwood, Long Branch           Pizza Place   
2   Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                   Bayview Village    Chinese Restaurant   
4                 Bedford Park, Lawrence Manor East           Coffee Shop   
..                                              ...                   ...   
92                      Willowdale, Willowdale West              Pharmacy   
93                                           Woburn           Coffee Shop   
94                                 Woodbine Heights           Curling Ice   
95                                  York Mills West                  Park   
96                         York Mills, Silver Hills             Cafeteria   

        2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0   Latin American Restaurant    Chinese Restaurant        Breakfast Spot   
1                    Pharmacy                   Gym                   Pub   
2                        Bank         Shopping Mall              Pharmacy   
3                        Bank   Japanese Restaurant                  Café   
4              Sandwich Place    Italian Restaurant       Thai Restaurant   
..                        ...                   ...                   ...   
92              Grocery Store               Butcher           Pizza Place   
93      Korean BBQ Restaurant      Cuban Restaurant            Donut Shop   
94                   Bus Stop                  Park            Beer Store   
95          Convenience Store           Yoga Studio      Cuban Restaurant   
96                Yoga Studio              Creperie            Donut Shop   

      5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0               Curling Ice            Donut Shop               Dog Run   
1               Coffee Shop             Gastropub     Convenience Store   
2            Ice Cream Shop           Supermarket            Restaurant   
3               Yoga Studio           Curling Ice            Donut Shop   
4   Comfort Food Restaurant             Pet Store              Pharmacy   
..                      ...                   ...                   ...   
92              Coffee Shop           College Gym      Cuban Restaurant   
93                  Dog Run   Distribution Center        Discount Store   
94             Skating Rink          Intersection    Athletics & Sports   
95               Donut Shop               Dog Run   Distribution Center   
96                  Dog Run   Distribution Center        Discount Store   

        8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0         Distribution Center        Discount Store                  Diner  
1                       Diner    Dim Sum Restaurant           Dessert Shop  
2   Middle Eastern Restaurant        Sandwich Place          Deli / Bodega  
3                     Dog Run   Distribution Center         Discount Store  
4                 Pizza Place                  Café                    Pub  
..                        ...                   ...                    ...  
92                    Dog Run   Distribution Center         Discount Store  
93                      Diner    Dim Sum Restaurant           Dessert Shop  
94                    Dog Run   Distribution Center         Discount Store  
95             Discount Store                 Diner     Dim Sum Restaurant  
96                      Diner    Dim Sum Restaurant           Dessert Shop  

[97 rows x 11 columns]

<b> Importing the required library for K-Means </b>

In [31]:
from sklearn.cluster import KMeans

<b> Running the K-Means algorithm on the dataframe </b>

In [32]:
k = 5
toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

kmeans = KMeans(n_clusters = k, random_state = 0)
kmeans.fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

<b> Checking the labels of the model </b>

In [33]:
kmeans.labels_

array([2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 0, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 0, 2, 2, 2, 2, 2,
       3, 2, 2, 2, 2, 2, 0, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2, 3, 2, 2, 2, 3,
       2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 0, 2, 2,
       3, 2, 3, 2, 2, 2, 2, 3, 4], dtype=int32)

<b> Adding clustering label to the dataframe </b>

In [34]:
neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

<b> Joining neighbourhoods_venues_sorted and df on neighbourhood to prepare for plotting </b>

In [35]:
toronto_merged = df.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on = 'Neighbourhood')
toronto_merged

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  \
0                                            Parkwoods  43.753259 -79.329656   
1                                     Victoria Village  43.725882 -79.315572   
2                            Regent Park, Harbourfront  43.654260 -79.360636   
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763   
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494   
..                                                 ...        ...        ...   
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944   
99                                Church and Wellesley  43.665860 -79.383160   
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558   
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509   
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999   

     Cluster Labels 1st Most Common Venue       2nd Most Common Venue  \
0               3.0                  Park           Food & Drink Shop   
1               0.0           Pizza Place                 Coffee Shop   
2               2.0           Coffee Shop                        Park   
3               2.0        Clothing Store           Accessories Store   
4               2.0           Coffee Shop            Sushi Restaurant   
..              ...                   ...                         ...   
98              2.0                  Pool                       River   
99              2.0      Sushi Restaurant                 Coffee Shop   
100             2.0    Light Rail Station                 Yoga Studio   
101             2.0        Baseball Field  Construction & Landscaping   
102             2.0           Social Club              Hardware Store   

      3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
0               Yoga Studio              Creperie                Dog Run   
1              Intersection          Hockey Arena  Portuguese Restaurant   
2                    Bakery               Theater         Breakfast Spot   
3    Furniture / Home Store              Boutique            Event Space   
4            Discount Store                  Bank     Italian Restaurant   
..                      ...                   ...                    ...   
98              Yoga Studio       Coworking Space    Distribution Center   
99                    Diner              Beer Bar             Hobby Shop   
100              Comic Shop            Skate Park                Brewery   
101             Yoga Studio      Cuban Restaurant             Donut Shop   
102          Discount Store        Sandwich Place           Burger Joint   

     6th Most Common Venue  7th Most Common Venue 8th Most Common Venue  \
0      Distribution Center         Discount Store                 Diner   
1              Yoga Studio       Department Store           Curling Ice   
2    Performing Arts Venue         Chocolate Shop                   Pub   
3              Coffee Shop  Vietnamese Restaurant               Gay Bar   
4                 Beer Bar    Japanese Restaurant             Nightclub   
..                     ...                    ...                   ...   
98          Discount Store                  Diner    Dim Sum Restaurant   
99                Creperie   Ethiopian Restaurant                   Pub   
100          Burrito Place         Farmers Market  Fast Food Restaurant   
101                Dog Run    Distribution Center        D

<b> Dropping al NaN values to prevent data skew </b>

In [36]:
toronto_merged_nonan = toronto_merged.dropna(subset = ['Cluster Labels'])

<b> Importing required libraries for plotting the clusters on map </b>

In [37]:
import matplotlib.cm as cm
import matplotlib.colors as colors

<b> Plotting the clusters on map </b>

In [38]:
map_clusters = folium.Map(location = [latitude, longitude], zoom_start = 11)

x = np.arange(k)
ys = [i + x + (i * x) ** 2 for i in range(k)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

markers_colors = []
for lat, long, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], 
                                   toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster' + str(int(cluster)) + '\n' + str(poi), parse_html = True)
    folium.CircleMarker([lat, long], radius = 5, popup = label, color = rainbow[int(cluster - 1)], 
                        fill = True, fill_color = rainbow[int(cluster - 1)]).add_to(map_clusters)

map_clusters

<b> We have successfully clustered Toronto neighbourhoods based on the venue categories </b>